component/discovery/discovery.go

```diff
@@ -0,0 +1,71 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/agent/component/metrics/scrape"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// MaxUpdateFrequency is the minimum time to wait between updating targets.
+// Currently not settable, since prometheus uses a static threshold, but
+// we could reconsider later.
+const MaxUpdateFrequency = 5 * time.Second
+
+// RunDiscovery is a utility for consuming and forwarding target groups from a discoverer.
+// It will handle collating targets (and clearing), as well as time based throttling of updates.
+// f should be a function that updates the component's exports, most likely calling `opts.OnStateChange()`.
+func RunDiscovery(ctx context.Context, d discovery.Discoverer, f func([]scrape.Target)) {
+	// all targets we have seen so far
+	cache := map[string]*targetgroup.Group{}
+
+	ch := make(chan []*targetgroup.Group)
+	go d.Run(ctx, ch)
+
+	// function to convert and send targets in format scraper expects
+	send := func() {
+		allTargets := []scrape.Target{}
+		for _, group := range cache {
+			for _, target := range group.Targets {
+				labels := map[string]string{}
+				for k, v := range group.Labels {
```

Comment:
nit: can you add a comment here explaining that it's important for the group to go first since the labels from target take precedence over the group labels?